text/0000-delegation-of-implementation.md

```diff
@@ -0,0 +1,235 @@
+    }
+}
+```
+Here we have no trait to delegate but the same method signatures are reused and semantically the situation is close to a trait-based implementation. A simple possibility could be to introduce a new trait. An alternative is to allow delegation at method level.
+```rust
+impl<'t, 'a,'tcx> fn node_ty for MemCategorizationContext<'t, 'a, 'tcx> use self.typer;
+```
+
+## More complex delegation
+
+`Self` can also appear inside more complex parameter/result types like `Option<Self>` or `&[Self]`. If we had HKT in Rust a partial solution based on [functor types][functors] might have been possible. It could still be possible to handle specific cases like precisely options and slices but I have not thought hard about it. The complexity might not be worth it.
+
+[functors]: https://wiki.haskell.org/Functor
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It creates some implicit code reuse. This is an intended feature but it could also be considered as dangerous. Modifying traits and surrogate types may automatically import new methods in delegating types with no compiler warning even in cases it is not appropriate (but this issue is the same as modifying a superclass in OOP).
+* The benefit may be considered limited for one-method traits.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## OOP inheritance
+
+As mentioned before, inheritance can handle similar cases with the advantage its concepts and mechanisms are well known. But with some drawbacks:
+* Multiple inheritance is possible but to my knowledge no serious proposition has been made for Rust and I doubt anyone wants to end up with a system as complex and tricky as C++ inheritance (whereas delegation is naturally "multiple delegation")
+* As said before inheritance mixes orthogonal concepts (code reuse and subtyping) and does not allow fine grain control over which part of the superclass interface is inherited.
+
+## Multiple derefs
+
+Some people noticed a similarity with trait `Deref`. A main limitation is that you can only deref to a single type. However one could imagine implementing multiple derefs by providing the target type as a generic parameter (`Deref<A>`) rather than as an associated type. But again you can find limitations:
+* As for inheritance visibility control is impossible: if `B` can be derefed to `A` then the entire public interface of `A` is accessible.
+* `Deref` only offers a superficial similarity. If `A` implements trait `Tr`, instances of `B` can sometimes be used where `Tr` is expected but as a counter example a `[B]` array is not assignable to `fn f(t: [T]) where T : Tr`. Derefs do not interact nicely with bounded generic parameters.
+
+## Compiler plugin
+
+I was suggested to write a compiler plugin. But I was also told that [type information is not accessible][type_information] (unless you can annotate the surrogate type yourself, which implies you must own it). Moreover I'm not sure a plugin could easily solve the partial delegation cases.
```

Comment:
Unfortunately, this would require two macro expansion phases (one for macros that can add new namable items and one for macros that can lookup named items).